### DISCUSSION TITLE
OneSignal

### DISCUSSION THREAD

**Discussion by nubo94** [vercel/next.js#17097] [Help]
Hello nextjs community!

Is there an example of integration with respect to [onesignal](https://onesignal.com/)?

I am currently trying to integrate oneSignal, but this service is not instantiated correctly, is there a correct way to start this service in nextjs?

Here is an extract of what I am trying to do.

```
import React, { useEffect } from "react";

function OneSignal() {
  useEffect(() => {
    if (process.browser) {
      const OneSignal = window.OneSignal || [];
      OneSignal.push(function () {
        OneSignal.init({
          appId: "4dc9ab15-03c1-459b-83a6-***********",
          allowLocalhostAsSecureOrigin: false,
          autoResubscribe: false,
          notifyButton: {
            enable: false,
          },
          welcomeNotification: {
            title: "...",
            message: "...",
            // "url": "" /* Leave commented for the notification to not open a window on Chrome and Firefox (on Safari, it opens to your webpage) */
          },
        });
        OneSignal.showNativePrompt();
      });
    }
  }, [process.browser]);

  return <></>;
}

export default OneSignal;
```

**Comment by M1ck0** [comment]
Why don't you use their official package from npm?

You can find it [here](https://www.npmjs.com/package/onesignal-node)

I am sure that you won't have those problems and it will be much easier to integrate.

**Reply by nubo94** [reply]
Thanks for your reply, I'll take a look!

**Reply by feedsbrain** [reply]
@nubo94 have you figure out how to init one signal with nextjs?

**Reply by M1ck0** [reply]
@feedsbrain 

Use OneSignal [official npm package](https://www.npmjs.com/package/onesignal-node).

If you have a problem with SSR on NextJS then import it on the client only. Take a look at NextJS [dynamic](https://nextjs.org/docs/advanced-features/dynamic-import) for that

**Comment by feedsbrain** [comment]
@M1ck0 [onesignal-node](https://www.npmjs.com/package/onesignal-node) has a compatibility issue with the latest LTS node. [react-onesignal](https://www.npmjs.com/package/react-onesignal) initialization looks similar to the official example from one signal. Although I don't get any error and my nextjs project successfully build using react-onesignal, it doesn't seem to register or subscribe to the service.

**Comment by feedsbrain** [comment]
Using dynamic as you mentioned @M1ck0, I can make `react-onesignal` works ... thanks for the tips!

```
 const initOneSignal = async () => {
    const OneSignal = (await import('react-onesignal')).default
    OneSignal.initialize('App ID', options)
  }

  useEffect(() => {
    initOneSignal()
  }, [])
```

**Reply by M1ck0** [reply]
Great. I am glad I could help

**Reply by scottkz** [reply]
> Great. I am glad I could help

@M1ck0 Maybe you can help me as well?

Hi everyone,

I am also using the same technique, but it's giving me a Service Worker 404 error. Is there a step that I'm just not doing here?

**NotifictionsHandler.js**
<pre>
  <code>
      import { useEffect } from 'react'
      import configs from '../../configs'

      export default function NotificationsHandler({ children }) {
        const initOneSignal = async () => {
        const OneSignal = (await import('react-onesignal')).default
        OneSignal.initialize(configs.onesignal.appId)
        }

       useEffect(() => {
         initOneSignal()
        }, [])

       return children
      }
  </code>
</pre>

**_app.jsx**
<pre>
  <code>
      import NotificationsHandler from '../modules/Auth/NotificationsHandler'
      
      export default class MyApp extends NextApp {
        componentDidMount() {
        this.setState({ mounted: true })
      }

      render() {
        const { mounted } = this.state || {}
        const { Component, pageProps } = this.props

        if (!mounted) {
          return null
        }

        return (
          ServiceProviders
            Component {...pageProps} 
            AppLoaderController 
            ModalRouter routes={modalRoutes} 
          ServiceProviders
        )
      }
    }

    function ServiceProviders({ children }) {
      return (
        NotificationsHandler
          CensoredUserGuard
             EnumsProvider{children}EnumsProvider
          CensoredUserGuard
        NotificationsHandler
              
      )
    }
  </code>
</pre>

Obviously leaving <> tags out so components will render, here, in my question. I tried downloading the OneSignalServiceWorker.js files and putting them in both the root of the Next.js project and under the pages directory, but no dice. How do you get ServiceWorkers to work with web on Next.js? I asked OneSignal and they seem clueless. Just kept sending me to React docs.

**Reply by M1ck0** [reply]
You need to put service worker in `public` directory of your `NextJS` project.

**Reply by scottkz** [reply]
@M1ck0 I very much appreciate your response. I assumed it was a missing step and nothing more. Thank you for helping me with that solution. 

One last question, to add tags to the user/device with the setTag/setTags methods, is that something that we just can’t do because of the dynamic imports? Adding the setTags method, for example directly under the .initialize() method or in the useEffect() block both give an OneSignal not initialized correctly error.

<pre>
  <code>
    export default function    NotificationsHandler({ children }) {
        const initOneSignal = async () => {
        const OneSignal = (await     import('react-onesignal')).default
        OneSignal.initialize(configs.onesignal.appId)
OneSignal.setTags(key: key, value: value)
        }

       useEffect(() => {
         initOneSignal()
        }, [])
</code>
 </pre>

**Reply by M1ck0** [reply]
I think that you are importing it wrong. You should import it on top of the file like all other imports using `next/dynamic` like this

```js
const OneSignal = dynamic(() => import('react-onesignal'), {
  ssr: false
});
```

And then you can maybe do something like
```js
// check if imported
if(OneSignal) {
  // do something if OneSignal is initalized
}
```

**Reply by scottkz** [reply]
Here' the whole file as it currently is...

```javascript
import React, { useEffect } from 'react'
import OneSignalContext from './OneSignalContext'
// import useAuth from '@braille-army/modules/auth/useAuth'
// import useUserSubscription from '../../Members/useUserSubscription'

export default function NotificationsHandler({ config, children }) {
  const [error, setError] = React.useState()
  // const auth = useAuth()
  // const { user } = auth
  console.log(error)
  // const userSubscription = useUserSubscription()
  // const { subscription } = userSubscription

  // Init initial OneSignal notification prompt
  const initOneSignal = async (username, hasSubscription) => {
    const OneSignal = (await import('react-onesignal')).default
    OneSignal.initialize(config.appId, {
      safari_web_id: 'web.braillearmy.com',
      notifyButton: {
        enable: true,
      },
    })
    // const keyValues = {
    //   user_username: username,
    //   braille_army_subscription: hasSubscription,
    // }
    // OneSignal.sendTags(keyValues)
  }

  useEffect(() => {
    // const username = user ? user.username : ''
    // const hasSubscription = !!(user && subscription)

    try {
      initOneSignal()
    } catch (error) {
      setError(error)
    }

    console.log(
      `OneSignal push notifications initiated
via appId [${config.appId.substring(0, 6)}...]`
    )
  }, [])

  return <OneSignalContext.Provider>{children}</OneSignalContext.Provider>
}

```

Everything is working correctly as far as OneSIgnal's ability to message users, but I cannot apply any of OneSignals methods, like .setTags() to be able to tell if each user has an account, and/or is part of our membership.

If I attempt to import it dynamically like you stated above I get an Unhandled Runtime Error
```TypeError: undefined is not a function (near '...OneSignal.initialize...')```

Trying to do this...
```javascript
import React from 'react'
import OneSignalContext from './OneSignalContext'
// import useAuth from '@braille-army/modules/auth/useAuth'
// import useUserSubscription from '../../Members/useUserSubscription'

import dynamic from 'next/dynamic'
const OneSignal = dynamic(() => import('react-onesignal'), {
  ssr: false,
})

export default function NotificationsHandler({ config, children }) {
  if (OneSignal) {
    OneSignal.initialize(config.appId, {
      safari_web_id: 'web.braillearmy.com',
      notifyButton: {
        enable: true,
      },
    })

    OneSignal.setTags(keyValues)
  }

  console.log(
    `OneSignal instantiated via appId [${config.appId.substring(0, 6)}...]`
  )

  return <OneSignalContext.Provider>{children}</OneSignalContext.Provider>
}
```

Once I import it dynamically it's treating OneSignal as a component now and not a call to the OneSignal library. Maybe I'm just missing the idea of dynamically importing it?

**Reply by M1ck0** [reply]
Oh my. This is my mistake. NextJS dynamic import does not support functions. Only components for now.

```js
const OneSignal = (await import('react-onesignal')).default
```
I am not an expert but why do you use `.default` here?

I have 1 thing on my mind as a last resort.

Have you tried loading `OneSignal` like this

```js
<script src="https://cdn.onesignal.com/sdks/OneSignalSDK.js" async=""></script>
```

You put this inside of `pages/_document.js` file. Read more about it [here](https://nextjs.org/docs/advanced-features/custom-document)

And then you might be able to use something this (from OneSignal's docs)
```js
var OneSignal = window.OneSignal || [];
```

And you could also check if `window` exists with this

```js
let OneSignal = []

if (typeof window !== "undefined") {
  OneSignal = window.OneSignal || [];
}
```

**Comment by feedsbrain** [comment]
So, I ended up doing this way:

Create `ReactOneSignal.tsx` component:
``` tsx
import React from 'react'

class ReactOneSignal extends React.Component {
  oneSignal = () => {
    return {
      __html: `window.OneSignal = window.OneSignal || []; OneSignal.push(function() { OneSignal.init({ appId: "${process.env.NEXT_PUBLIC_ONESIGNAL_APP_ID}" }); });`
    }
  }

  render = (): React.ReactNode => {
    return (
      <>
        <script src="https://cdn.onesignal.com/sdks/OneSignalSDK.js" async />
        <script dangerouslySetInnerHTML={this.oneSignal()} />
      </>
    )
  }
}

export default ReactOneSignal
```

And then use that component in `_document.tsx`:
```tsx
import Document, { Html, Head, Main, NextScript } from 'next/document'
import ReactOneSignal from '../components/ReactOneSignal'

class CustomDocument extends Document {
  static getInitialProps = async (ctx) => {
    const initialProps = await Document.getInitialProps(ctx)
    return { ...initialProps }
  }

  render = () => {
    return (
      <Html>
        <Head >
          <link rel="shortcut icon" href="/favicon.png" />
          <ReactOneSignal />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default CustomDocument 
```

**Reply by Inzendis** [reply]
> So, I ended up doing this way:
> 
> Create `ReactOneSignal.tsx` component:
> 
> ```tsx
> import React from 'react'
> 
> class ReactOneSignal extends React.Component {
>   oneSignal = () => {
>     return {
>       __html: `window.OneSignal = window.OneSignal || []; OneSignal.push(function() { OneSignal.init({ appId: "${process.env.NEXT_PUBLIC_ONESIGNAL_APP_ID}" }); });`
>     }
>   }
> 
>   render = (): React.ReactNode => {
>     return (
>       <>
>         <script src="https://cdn.onesignal.com/sdks/OneSignalSDK.js" async />
>         <script dangerouslySetInnerHTML={this.oneSignal()} />
>       </>
>     )
>   }
> }
> 
> export default ReactOneSignal
> ```
> 
> And then use that component in `_document.tsx`:
> 
> ```tsx
> import Document, { Html, Head, Main, NextScript } from 'next/document'
> import ReactOneSignal from '../components/ReactOneSignal'
> 
> class CustomDocument extends Document {
>   static getInitialProps = async (ctx) => {
>     const initialProps = await Document.getInitialProps(ctx)
>     return { ...initialProps }
>   }
> 
>   render = () => {
>     return (
>       <Html>
>         <Head >
>           <link rel="shortcut icon" href="/favicon.png" />
>           <ReactOneSignal />
>         </Head>
>         <body>
>           <Main />
>           <NextScript />
>         </body>
>       </Html>
>     )
>   }
> }
> 
> export default CustomDocument 
> ```

Does this still work? Im trying to get it to work on Next JS and I have followed their documentation with is not accurate for Next JS. I used your example and it still doesn't work.

The only changes is that the cdn link now ends with /OneSignalSDKWorker.js and my public App ID.

**Reply by luoser** [reply]
looks like they had some breaking changes published a month ago, try upgrading to use their npm package? worked for me 

https://github.com/OneSignal/react-onesignal/blob/main/MigrationGuide.md
